### ISSUE TITLE
Duplicated function Raster.set_dtypes and Raster.astype and use of several dtypes

### ISSUE THREAD

**Issue by adehecq** [GlacioHack/GeoUtils#206]
It seems like `Raster.set_dtypes` and `Raster.astype` have the same functionality.
This is my fault as I didn't check for the existing functionality before adding 'astype'.

This raises two questions:
1) Which version should we keep?
set_dtypes seem to cover cases with multiple dtypes, but in the end, it simply uses the first value (see 2nd point). It has more checks, but in the end, those are already covered by numpy.astype, which raises a straightforward error if dtype is incorrect. It replaces the array.
astype seem much more succinct and easy to understand. The only drawback is that currently it does not enable to do in-place changes, although it would be an easy fix.
 
2) Most of the time, we ignore that dtypes can contain multiple values. Sometimes, we pretend to take it into account (but not really). See for example: 
- https://github.com/GlacioHack/GeoUtils/blob/059f6a3025d0d508d6135a6daffeb1c28ffe563b/geoutils/georaster.py#L807
- https://github.com/GlacioHack/GeoUtils/blob/059f6a3025d0d508d6135a6daffeb1c28ffe563b/geoutils/georaster.py#L1040
So the question is, should we actually try to cover the case where data types are different? 
If not, where should we make this clear?
rasterio is *supposed* to handle different data types, although it is impossible to create a raster with different dtypes using rasterio.open (https://rasterio.readthedocs.io/en/latest/api/rasterio.html?highlight=open#rasterio.open). 
Different dtypes are also not supported with GTIff: https://stackoverflow.com/questions/58761022/can-i-append-a-band-with-different-dtype-to-a-geotiff-using-rasterio

Any thoughts?

**Comment by adehecq** [comment]
Also, set_dtypes with option update_array=False does not return anything... It should return a Raster instead.

**Comment by adehecq** [comment]
And it does not preserve the mask...
```
import geoutils.georaster as gr
r = gr.Raster(datasets.get_path("landsat_B4"))
r.data.mask = np.zeros(r.data.shape)
r.data.mask[0, 0, 0] = True
r.set_dtypes(np.int8)
r.data
```
-> mask is gone!

**Comment by rhugonnet** [comment]
> Also, set_dtypes with option update_array=False does not return anything... It should return a Raster instead.

Are you critizing my superb implementation of 6 months ago? :grin: First weeks of geoutils, not sure we had the mask yet.

> So the question is, should we actually try to cover the case where data types are different?

I don't think we should. This is literally less than <0.01% of use cases.

> If not, where should we make this clear?

Yes! Somewhere in the docs, not too deep.

> set_dtypes with option update_array=False

I think rasterio can return an error otherwise, so I commented in the function that it should be always true.

> Which version should we keep?

Looking at `astype`, it seems that it will only change the content of the array using `from_array()`, but not the underlying metadata value `.dtype` of the rasterio dataset/Raster object? It is creating the Memory object by calling again the `.dtype` of the Raster object, so I don't understand how it will work to change the dtype of the object (let's say from float32 to int16)?

**Comment by adehecq** [comment]
> > Also, set_dtypes with option update_array=False does not return anything... It should return a Raster instead.
> 
> Are you critizing my superb implementation of 6 months ago? grin First weeks of geoutils, not sure we had the mask yet.
> 
Haha exactly! I wasn't sure, but I had in mind that you wrote it ;-)

> > So the question is, should we actually try to cover the case where data types are different?
> 
> I don't think we should. This is literally less than <0.01% of use cases.

That's also what I think.
> 
> > If not, where should we make this clear?
> 
> Yes! Somewhere in the docs, not too deep.

At least, I'll make it explicit in the function.

> 
> > set_dtypes with option update_array=False does not return anything..
> 
> I think rasterio can return an error otherwise, so I commented in the function that it should be always true.
> EDIT: this is clearly an issue, we need to capture rasterio's error and say that just updating the dtype does not match the array, or something.

Maybe it didn't exist back then, but it can be done using from_array. That's what is done in astype.

> 
> > Which version should we keep?
> 
> Looking at `astype`, it seems that it will only change the content of the array using `from_array()`, but not the underlying metadata value `.dtype` of the rasterio dataset/Raster object? It is creating the Memory object by calling again the `.dtype` of the Raster object, so I don't understand how it will work to change the dtype of the object (let's say from float32 to int16)?

That's the magic of my implementation! ;-) So simple and yet it works. Just test it!
```
import geoutils.georaster as gr
from geoutils import datasets
r = gr.Raster(datasets.get_path("landsat_B4"))
rout = r.astype(np.float32)
print(rout.dtypes)
```
(well, ok it's the magic of from_array behind it...)

**Comment by rhugonnet** [comment]
Ah yes, I get it now! `from_array()` uses the np.ndarray and not the Raster attribute.
I also remember why started writing set_dtype this way: to possibly avoid recreating a Raster in-memory: in place modification, as you said. Only redefining the metadata, and not the array. 
But I'm not sure this is super useful, or that it really works this way for now anyway.

**Comment by adehecq** [comment]
Solved by #208.
